File: task1/program (variant 13, 19)/plant.cpp
//
// Функции обработки растения.
//

#include "plant.h"

// Ввод параметров растения из указанного потока.
plant* Input(FILE *in_file) {
    plant *new_plant;
    int plant_type;
    new_plant = new plant;

    fscanf(in_file, "%i", &plant_type);
    fscanf(in_file, "%s", new_plant -> name);
    switch(plant_type) {
        case(1):
            new_plant -> plant_type = plant::TREE;
            Input(new_plant -> un_tree, in_file);
            return new_plant;
        case(2):
            new_plant -> plant_type = plant::BUSH;
            Input(new_plant -> un_bush, in_file);
            return new_plant;
        case(3):
            new_plant -> plant_type = plant::FLOWER;
            Input(new_plant -> un_flower, in_file);
            return new_plant;
        default:
            return 0;
    }
}

// Случайная генерация параметров растения.
plant *RandomInput() {
    plant *new_plant = new plant;
    int plant_type_number = rand() % 3 + 1;

     GenerateRandomName(new_plant -> name);
    switch(plant_type_number) {
        case(1):
            new_plant -> plant_type = plant::TREE;
            RandomInput(new_plant -> un_tree);

            break;
        case(2):
            new_plant -> plant_type = plant::BUSH;
            RandomInput(new_plant->un_bush);

            break;
        case(3):
            new_plant->plant_type = plant::FLOWER;
            RandomInput(new_plant->un_flower);

            break;
        default:
            return 0;
    }

    return new_plant;
}

// Вывод параметров растения в указанный поток.
void Output(plant &out_plant, FILE *out_file) {
    switch(out_plant.plant_type) {
        case (plant::TREE):
            Output(out_plant.un_tree, out_plant.name, out_file);
            break;
        case (plant::FLOWER):
            Output(out_plant.un_flower, out_plant.name, out_file);
            break;
        case (plant::BUSH):
            Output(out_plant.un_bush, out_plant.name, out_file);
            break;
        default:
            fprintf(out_file, "Incorrect plant.\n");
    }
}

// Генерация случайного имени.
void GenerateRandomName(char *name) {
    const int name_length = rand() % 20 + 1;

    for (int i = 0; i < name_length; ++i) {
        name[i] = static_cast<char>(rand() % 26 + 97);
    }
    name[name_length] = '\0';
}

// Поиск частного от деления числам гласных букв в названии на общую длину названия.
double CalculateRatioOfVowelsToAllLetters(plant &plant) {
    int count = std::strlen(plant.name);
    int vowels_count = 0;
    char vowels[] = "aoiue";

    for (int i = 0; i < std::strlen(plant.name); ++i) {
        for (int vowel_index = 0; vowel_index < 5; ++vowel_index) {
            if (vowels[vowel_index] == plant.name[i]) {
                ++vowels_count;
                break;
            }
        }
    }
    return static_cast<double>(vowels_count) / count;
}

File: task3/program(variant 13, 19)/plants/plant.py
from typing.io import TextIO
from random import randint
from decimal import Decimal
from abc import ABC, abstractmethod


class Plant(ABC):
    """Функции для обработки растения."""

    def __init__(self):
        """Инициализация параметров растения."""
        self._name = ""

    @abstractmethod
    def input(self, plant_info: str):
        """Считывание параметров растения - абстрактный метод."""
        pass

    @abstractmethod
    def output(self, out_file: TextIO):
        """Вывод параметров растения - абстрактный метод."""
        pass

    @abstractmethod
    def random_input(self):
        """Рандомная генерация параметров растения - абстрактный метод."""
        pass

    @staticmethod
    def generate_random_name() -> str:
        """Генерация рандомного имени."""
        name_length = randint(1, 20)
        name = ""

        for i in range(name_length):
            name += chr(randint(97, 122))
        return name

    def calculate_ratio_of_vowels_to_all_letters(self) -> Decimal:
        """Поиск частного от деления числам гласных букв в названии на общую длину названия."""
        vowels_count = 0
        vowels = "aoiue"

        for char in self._name:
            if char in vowels:
                vowels_count += 1

        return Decimal(vowels_count) / Decimal(len(self._name))
